From ef379eae8534a7bc4847d8545573b00c5f4887ae Mon Sep 17 00:00:00 2001
From: Ralf Lici <ralf@mandelbit.com>
Date: Thu, 14 May 2026 12:45:17 +0200
Subject: [PATCH] ovpn-dco: serialize peer NAPI add/delete

Older kernels do not internally serialize updates to netdev->napi_list.
Peer creation runs from netlink while peer release can run asynchronously,
so serialize per-peer NAPI registration and removal with an instance mutex.

Upstream-Status: Backport [ef379eae8534a7bc4847d8545573b00c5f4887ae]
Signed-off-by: Ralf Lici <ralf@mandelbit.com>
Signed-off-by: Antonio Quartulli <antonio@openvpn.net>
---
 drivers/net/ovpn-dco/ovpn.c       |  1 +
 drivers/net/ovpn-dco/ovpnstruct.h |  4 ++++
 drivers/net/ovpn-dco/peer.c       | 24 +++++++++++++++++++-----
 3 files changed, 24 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ovpn-dco/ovpn.c b/drivers/net/ovpn-dco/ovpn.c
index 82e9f94..0961c59 100644
--- a/drivers/net/ovpn-dco/ovpn.c
+++ b/drivers/net/ovpn-dco/ovpn.c
@@ -66,6 +66,7 @@ int ovpn_struct_init(struct net_device *dev)
 
 	spin_lock_init(&ovpn->lock);
 	spin_lock_init(&ovpn->peers.lock);
+	mutex_init(&ovpn->napi_lock);
 
 	ovpn->crypto_wq = alloc_workqueue("ovpn-crypto-wq-%s",
 					  WQ_CPU_INTENSIVE | WQ_MEM_RECLAIM, 0,
diff --git a/drivers/net/ovpn-dco/ovpnstruct.h b/drivers/net/ovpn-dco/ovpnstruct.h
index 98f0953..bcedf8b 100644
--- a/drivers/net/ovpn-dco/ovpnstruct.h
+++ b/drivers/net/ovpn-dco/ovpnstruct.h
@@ -13,6 +13,7 @@
 #include "peer.h"
 
 #include <uapi/linux/ovpn_dco.h>
+#include <linux/mutex.h>
 #include <linux/spinlock.h>
 #include <linux/workqueue.h>
 
@@ -27,6 +28,9 @@ struct ovpn_struct {
 	/* protect writing to the ovpn_struct object */
 	spinlock_t lock;
 
+	/* protect NAPI add/delete operations */
+	struct mutex napi_lock;
+
 	/* workqueue used to schedule crypto work that may sleep */
 	struct workqueue_struct *crypto_wq;
 	/* workqueue used to schedule generic event that may sleep or that need
diff --git a/drivers/net/ovpn-dco/peer.c b/drivers/net/ovpn-dco/peer.c
index 167fa62..14a5454 100644
--- a/drivers/net/ovpn-dco/peer.c
+++ b/drivers/net/ovpn-dco/peer.c
@@ -36,6 +36,23 @@ static void ovpn_peer_expire(struct timer_list *t)
 	ovpn_peer_del(peer, OVPN_DEL_PEER_REASON_EXPIRED);
 }
 
+static void ovpn_peer_napi_add(struct ovpn_peer *peer)
+{
+	mutex_lock(&peer->ovpn->napi_lock);
+	netif_napi_add_tx_weight(peer->ovpn->dev, &peer->napi, ovpn_napi_poll,
+				 NAPI_POLL_WEIGHT);
+	napi_enable(&peer->napi);
+	mutex_unlock(&peer->ovpn->napi_lock);
+}
+
+static void ovpn_peer_napi_del(struct ovpn_peer *peer)
+{
+	mutex_lock(&peer->ovpn->napi_lock);
+	napi_disable(&peer->napi);
+	netif_napi_del(&peer->napi);
+	mutex_unlock(&peer->ovpn->napi_lock);
+}
+
 /* Construct a new peer */
 static struct ovpn_peer *ovpn_peer_create(struct ovpn_struct *ovpn, u32 id)
 {
@@ -89,9 +106,7 @@ static struct ovpn_peer *ovpn_peer_create(struct ovpn_struct *ovpn, u32 id)
 	}
 
 	/* configure and start NAPI */
-	netif_napi_add_tx_weight(ovpn->dev, &peer->napi, ovpn_napi_poll,
-				 NAPI_POLL_WEIGHT);
-	napi_enable(&peer->napi);
+	ovpn_peer_napi_add(peer);
 
 	dev_hold(ovpn->dev);
 
@@ -224,8 +239,7 @@ static void ovpn_peer_release_rcu(struct rcu_head *head)
 
 void ovpn_peer_release(struct ovpn_peer *peer)
 {
-	napi_disable(&peer->napi);
-	netif_napi_del(&peer->napi);
+	ovpn_peer_napi_del(peer);
 
 	if (peer->sock)
 		ovpn_socket_put(peer->sock);
-- 
2.34.1

